rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,335 @@
+   re-submission. This MAY be used to penalize jobs which have been running for
+   too long, or jobs which have been repeatedly failing. Note that both the
+   user and a worker can request a job to be re-submitted.
+
+```
+update(Tx, Type, JobId, JobOpts, WorkerLockId) -> ok | worker_conflict | canceled
+```
+ - This MAY be called to update a job's progress (update sequence, number of
+   changes left, etc.) This function MUST be called at least as often as the
+   `ActivityTimeout` in order for the activity monitor to not re-submit the
+   job into the pending queue due to inactivity.
+
+When the functions above return a `worker_conflict` value it means the activity
+monitor has already re-enqueued the job. It is now either pending or is executed
+by another worker. In that case, the caller MUST stop running the job, and they
+MUST NOT update the job's entry any longer.
+
+If `canceled` is returned, it means the user has requested the job to stop
+executing. In that case the worker MUST top running the job.
+
+#### Mutual Exclusion
+
+If each worker updates their job status on time, and the activity monitor is
+running correctly, the same job SHOULD be executed by no more than one worker
+at a time. But, if the worker process is blocked for too long (for instance, in
+an overload scenario), it may fail to update its status often enough, and the
+activity monitor MAY re-enqueue the job. However, even in such cases, the
+mutual exclusion constraint can be maintained as long as `update(Tx,...)`
+function is called in the same transaction as the type-specific DB writes, and
+its result is checked for the `worker_conflict` return.
+
+# Implementation Details
+
+## Data Model
+
+`("couch_jobs", "data", Type, JobId) = (Sequence, WorkerLockId, JobOpts)`
+`("couch_jobs", "pending", Type, Priority, JobId) = ""`
+`("couch_jobs", "watches", Type) = Sequence`
+`("couch_jobs", "activity_timeout", Type) = ActivityTimeout`
+`("couch_jobs", "activity", Type, Sequence) = JobId`
+
+`Type`, `Priority`, and `JobId` MUST be any type supported by the FoundationDB
+tuple layer.
+
+`JobOpts` MUST be a JSON object encoded as a string. It MAY have `"priority"`,
+`"cancel"`, `"resubmit"`, or `"data"` fields.
+
+## Job Lifecyle Implementation
+
+This section describes how the framework implements some of the API functions.
+
+`add()` :
+  * Add the new job to the main jobs table
+  * If a job with the same `JobId` exists, return an error
+  * `WorkerLockId` is set to `null`.
+
+`accept()` :
+ * Generate a unique `WorkerLockId` UUID.
+ * Attempt to dequeue the item from the pending queue, then assign it the
+   `WorkerLockId` in the jobs table.
+ * Create entry in the `"activity"` subspace.
+ * Update sequence in the `"watches"` subspace for that job's type.
+
+`update()`:
+ * Check if `WorkerLockId` matches, otherwise return `worker_conflict`
+ * Check if `"canceled"` field in the `JobOpts` is `true`, if so, return `canceled`
+ * Delete old `"activity"` sequence entry
```

Comment:
This is interesting, what are the reasons for using a sequence to track this versus just an integer that we could atomically increment for during each update?

Reply:
Good question. I should develop the algorithm for the activity monitor a bit more.

The idea is that if the jobs are updated on-time, the activity check will return almost no results and will not need to keep anything in memory except the "watches" sequence for the from the previous check (`ActivityTimeout` seconds ago). With entries being bumped with atomic counters, we'd have to keep the whole job table in memory to compare the counters. I think both approaches might work for the purposes of activity monitoring, and if we find that there are significant performance issues, we could switch to using atomic counters. 

A few more details:

Every time a job is updating its "activity", it bumps the "watches" sequence. That makes "watches" sequence point to the newest updated sequence for that job type. That allows anyone, in this case the activity monitor, to remember the "watches" sequence value, then return some time later and ask "has anything not been updated since time I saw that sequence" by querying any entries *less* than its remembered sequence. If everything is working properly and jobs are updating, it should get an empty list.

This scheme also allows efficient watching of new change since the last remembered sequence , by querying anything *greater* than the last seen "watches" sequence. So it allows relatively efficient monitoring of jobs without having a separate watch for each job entry.

Reply:
Ok cool. Could you add that info to the RFC that explains it really nicely.